buildSrc/subprojects/parser/src/main/java/org/apache/calcite/buildtools/parser/DialectTraverser.java

```diff
@@ -214,6 +228,27 @@ private void traverse(
     }
   }
 
+  /**
+   * Adds all of the elements in {@code otherMap} to {@code mainMap}. If a key
+   * is already present in {@code mainMap} it is removed before being updated.
+   * This is done to ensure that the {@code filePath} gets updated if a given keyword
+   * has been overridden.
+   *
+   * @param mainMap The map that is getting entries added to it
+   * @param otherMap The map whose entries are being added from
+   */
+  private void addOrReplaceEntries(Map<Keyword, String> mainMap,
```

Comment:
Can we just use mainMap.putAll(otherMap) for this?

Reply:
That was the original implementation but since filePath isn't used to compute hashCode or equals() only the value of the keyword would get replaced leaving the filePath to be not updated.

Reply:
Thanks for pointing this out. This might point to a larger issue in our usage of Keywords as keys.

Since Keyword doesn't implement hashCode() or equals() using all of its fields, it may not be a good contender as a key in a Set/Map. Keys in these data structures should match the intuition that if key1.equals(key2) is true, then the keys are identical.

Reply:
I created a possible alternative here: https://github.com/AndrewPochapsky/calcite/pull/1. I changed Map<Keyword, String> to Map<String, Keyword> with a key of keyword name. Let me know what you think :)

Reply:
Cool thanks for coding that up. LGTM.

Reply:
Great, so I should merge it into this branch?

Reply:
Yes, please do. And then we can review the version with those changes in this PR.